src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs

```diff
@@ -89,60 +89,57 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
         /// <inheritdoc />
         public async Task<string> SetupAsync(SetupContext context)
         {
+            string executionId = null;
             var initialState = context.ShellSettings.State;
             try
             {
-                var executionId = await SetupInternalAsync(context);
+                // Set the shell state to 'Initializing' to guarantee that any subsequent HTTP requests
+                // will result in a 'Service Unavailable' response while the tenant setup is in progress.
+                context.ShellSettings.AsInitializing();
 
-                if (context.Errors.Any())
+                executionId = await SetupInternalAsync(context);
+
+                if (context.Errors.Count > 0)
                 {
                     context.ShellSettings.State = initialState;
                     await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
                 }
-
-                return executionId;
             }
-            catch
+            catch (Exception e)
             {
+                _logger.LogError(e, "An error occurred while setting up the tenant.");
+                context.Errors.Add(String.Empty, S["An error occurred while setting up the tenants: {0}", e.Message]);
+
                 context.ShellSettings.State = initialState;
                 await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
-
-                throw;
             }
+
+            return executionId;
         }
 
         private async Task<string> SetupInternalAsync(SetupContext context)
         {
-            string executionId;
+            _logger.LogInformation("Running setup for tenant '{TenantName}'.", context.ShellSettings?.Name);
 
-            if (_logger.IsEnabled(LogLevel.Information))
```

Comment:
Why removing this check?

Reply:
Because LogInformation will only print to the file when the log level is info+. That check isn't needed

Reply:
But it will still parse and tokenise the string to insert serialized params.

Here not so important but that's a good practice.

Reply:
If this check is needed it should be part of the logger implementation not our code. Anyway, I'll restore it.

Reply:
It was suggested but, as I remember, at some point they didn't because too late, so to prevent double checking. Maybe in the meantime it has been added, here some interesting links.

https://andrewlock.net/exploring-dotnet-6-part-8-improving-logging-performance-with-source-generators/

https://github.com/dotnet/runtime/issues/63907

https://github.com/dotnet/runtime/issues/51927